File: 2020/day2.swift
import Foundation

// 3-4 j: tjjj

struct PasswordPolicy {
  let min: Int
  let max: Int
  let match: String
  let password: String

  var isValid: Bool {
    let array = Array(password)
    let a = String(array[min - 1])
    let b = String(array[max - 1])

    return (a == match || b == match) && a != b
  }
}

do {
  let content = try String(contentsOfFile: "input2.txt")
  let strings = content.components(separatedBy: "\n")

  let policies = strings.compactMap { element -> PasswordPolicy? in
    let components = element.components(separatedBy: " ")
    if components.count == 3 {
      let first = components.first?.components(separatedBy: "-")

      let min = Int(first!.first!)!
      let max = Int(first!.last!)!

      let match = String(components[1].first!)

      let password = components.last!

      return PasswordPolicy(min: min, max: max, match: match, password: password)
    }
    return nil
  }
  print(policies.filter {
    $0.isValid
  }.count)
}

File: 2020/day2.dart
import 'dart:io';

class PasswordPolicy {
  PasswordPolicy({
    this.min,
    this.max,
    this.match,
    this.password,
  });

  int min;
  int max;
  String match;
  String password;

  bool get isValid {
    final a = password[min - 1];
    final b = password[max - 1];

    return (a == match || b == match) && a != b;
  }

  @override
  String toString() {
    return '$min - $max - $match - $password \n';
  }
}

main() {
  var path = 'input2.txt';
  final file = new File(path);
  final List<String> list = file.readAsLinesSync();
  final List<PasswordPolicy> policies = list.map((e) {
    var components = e.split(' ');
    var minMax = components[0].split('-');
    return PasswordPolicy(
        min: int.parse(minMax[0]),
        max: int.parse(minMax[1]),
        match: components[1][0],
        password: components[2]);
  }).toList();

  print(policies.where((element) => element.isValid).length);
}
